### PR TITLE
Extract ResultTableManager from ResultManager

### PR DESCRIPTION
**Pull request by tgerdesnv**
Extracted ResultTableManager class out of ResultManager class, and then cleaned up ResultManager
The result is cleaner, but still far from "clean"

ResultTableManager:  Manages (creates and adds data to) ResultTables
- All code copied out of ResultManager
- Relevant testing code put into test_result_table_manager.py
- No new testing added

ResultManager: Holds and sort results
- Added public access functions:
  - get_model_names()
  - get_model_sorted_results()
  - get_across_model_sorted_results()
  - get_results()
  - get_server_only_data()
- Added basic testing for all of those new functions, as well as _find_default_configs_throughput()

Boyscouts:
- Fixed cases where other classes were reaching into ResultManager private variables
- Extracted evaluate_mock_config() into test_utils
- Extracted load_[single|multi]_result_manager() into test_utils

### CODE REVIEW

File: model_analyzer/result/result_manager.py [hunk 1]
```diff
@@ -465,430 +212,141 @@ def top_n_results(self, model_name=None, n=-1, include_default=False):
...
-                    gpu_fields, batch_sizes, concurrencies, satisfies,
-                    model_name, model_config_name, dynamic_batchings,
-                    instance_groups, max_batch_sizes)
-
-                self._add_uuid_to_gpu_row(gpu_row, gpu_uuid, gpu_fields)
-                self._add_metrics_to_gpu_row(gpu_row, metrics, gpu_fields)
-
-                self._result_tables[
-                    self.model_gpu_table_key].insert_row_by_index(row=gpu_row)
+        self._state_manager.set_state_variable('ResultManager.results',
```
Comment (tgerdesnv) [inline model_analyzer/result/result_manager.py:250]: This class is the only place we should ever use these hardcoded keys, and they should really only do it on one or two places (or use a const)

File: model_analyzer/result/result_table_manager.py [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+        # being collected.
+        gpu_specific_metrics_from_measurements = {}
+        non_gpu_specific_metrics_from_measurements = {}
+        # Find metrics if one or more of them is not provided
+        if gpu_specific_metrics == None or non_gpu_specific_metrics == None:
+            # Server data
+            data = self._result_manager.get_server_only_data()
+            for gpu_uuid, gpu_metrics in data.items():
+                for gpu_metric in gpu_metrics:
+                    if gpu_metric.tag not in gpu_specific_metrics_from_measurements:
```
Comment (tgerdesnv) [inline model_analyzer/result/result_table_manager.py:153]: Refactor functions in this file (especially this one)

File: model_analyzer/result/result_table_manager.py [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+        if include_title:
+            writer.write('\n'.join([
+                table.title() + ":",
+                table.to_formatted_string(separator=column_separator,
+                                          ignore_widths=ignore_widths), "\n"
+            ]))
+        else:
+            writer.write(
+                table.to_formatted_string(separator=column_separator,
+                                          ignore_widths=ignore_widths) + "\n\n")
```
Comment (debermudez) [inline model_analyzer/result/result_table_manager.py:340]: Is the extra new line intentional here?  
To refactor this method slightly, I would build the string and then make 1 call to the writer with the formatted string.

File: model_analyzer/result/result_table_manager.py [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+
+    def _add_result_table(self, table_key, title, headers):
+        """
+        Utility function that creates a table with column
+        headers corresponding to perf_analyzer arguments
+        and requested metrics. Also sets the result
+        comparator for that table.
+        """
+
+        # Create headers
```
Comment (debermudez) [inline model_analyzer/result/result_table_manager.py:385]: Is this comment correct?
This line creates the whole table including _headers_ right?

File: tests/common/test_utils.py [hunk 5]
```diff
@@ -28,6 +35,58 @@
...
+    ]
+    yaml_content = convert_to_bytes("""
+        analysis_models: resnet50_libtorch,vgg19_libtorch
+    """)
+    config = evaluate_mock_config(args, yaml_content, subcommand="analyze")
+    state_manager = AnalyzerStateManager(config=config, server=None)
+    state_manager.load_checkpoint(checkpoint_required=True)
+
+    result_manager = ResultManager(config=config, state_manager=state_manager)
+    return result_manager, config
```
Comment (nv-hwoo) [inline tests/common/test_utils.py:87]: I think we can further refactor the code by combining them and take arguments.

File: tests/test_result_manager.py [hunk 6]
```diff
@@ -14,251 +14,157 @@
...
+
+    def test_get_multi_model_names(self):
+        """ Test get_model_names for a multi-model run """
+        result_manager, _ = load_multi_model_result_manager()
+        result_manager.compile_and_sort_results()
+        self.assertEqual(result_manager.get_model_names(),
+                         ["resnet50_libtorch,vgg19_libtorch"])
+
+    def test_get_model_sorted_results_bad_name(self):
+        """ Test get_model_sorted_results will assert when called with name that doesn't exist """
```
Comment (nv-hwoo) [inline tests/test_result_manager.py:119]: Add newline for long docstrings.

File: tests/test_result_manager.py [hunk 7]
```diff
@@ -14,251 +14,157 @@
...
+
+        result_manager, _ = load_multi_model_result_manager()
+        result_manager.compile_and_sort_results()
+        with self.assertRaises(TritonModelAnalyzerException):
+            result_manager.get_model_sorted_results("SHOULD_ERROR")
+
+    def test_get_single_model_sorted_results(self):
+        """ Test get_model_sorted_results returns a valid ResultHeap with only results for that model """
```
Comment (nv-hwoo) [inline tests/test_result_manager.py:127]: Add newline for long docstrings.

File: tests/test_result_manager.py [hunk 8]
```diff
@@ -14,251 +14,157 @@
...
+        """ Test get_model_sorted_results returns a valid ResultHeap """
+        result_manager, _ = load_multi_model_result_manager()
+        result_manager.compile_and_sort_results()
+        sorted_results = result_manager.get_model_sorted_results(
+            "resnet50_libtorch,vgg19_libtorch")
+        self.assertTrue(isinstance(sorted_results, ResultHeap))
+        self.assertEqual(17, len(sorted_results.results()))
+
+    def test_get_across_model_sorted_results(self):
+        """ Test get_across_model_sorted_results returns a valid ResultHeap with all results across all models """
```
Comment (nv-hwoo) [inline tests/test_result_manager.py:145]: Add newline for long docstrings.

File: tests/test_model_manager.py [hunk 9]
```diff
@@ -15,21 +15,13 @@
 import unittest
 from .common import test_result_collector as trc
 
-from .common.test_utils import construct_run_config_measurement
-from .mocks.mock_config import MockConfig
+from .common.test_utils import construct_run_config_measurement, evaluate_mock_config
```
Comment (debermudez) [inline tests/test_model_manager.py:18]: Do you still need to import construct_run_config_measurement?

File: tests/test_perf_analyzer_config_generator.py [hunk 10]
```diff
@@ -13,16 +13,10 @@
 # limitations under the License.
 
 import unittest
-from tests.common.test_utils import construct_perf_analyzer_config
+from tests.common.test_utils import construct_perf_analyzer_config, evaluate_mock_config, convert_to_bytes, construct_run_config_measurement
```
Comment (debermudez) [inline tests/test_perf_analyzer_config_generator.py:16]: Do you need all of these imports?

File: model_analyzer/result/result_table_manager.py [hunk 11]
```diff
@@ -0,0 +1,609 @@
...
+
+        self._write_results(writer=FileWriter(), column_separator=' ')
+
+    def export_results(self):
+        """
+        Makes calls to _write_results out to streams or files. If
+        exporting results is requested, uses a FileWriter for specified output
+        files.
+        """
+
```
Comment (nv-hwoo) [inline model_analyzer/result/result_table_manager.py:96]: Remove space.

File: model_analyzer/result/result_table_manager.py [hunk 12]
```diff
@@ -0,0 +1,609 @@
...
+            result_heap = self._result_manager.get_model_sorted_results(model)
+
+            while not result_heap.empty():
+                self._tabulate_measurements(result_heap.next_best_result())
+
+    def write_results(self):
+        """
+        Writes table to console
+        """
+
```
Comment (nv-hwoo) [inline model_analyzer/result/result_table_manager.py:87]: Remove space.

File: tests/test_plot_manager.py [hunk 13]
```diff
@@ -15,20 +15,13 @@
 import unittest
 
 from .common import test_result_collector as trc
-from .common.test_utils import convert_to_bytes, default_encode, ROOT_DIR
-from .mocks.mock_config import MockConfig
-
-from model_analyzer.cli.cli import CLI
-from model_analyzer.config.input.config_command_analyze \
-    import ConfigCommandAnalyze
+from .common.test_utils import convert_to_bytes, evaluate_mock_config, ROOT_DIR
```
Comment (debermudez) [inline tests/test_plot_manager.py:18]: imports correct?

File: model_analyzer/result/result_table_manager.py [hunk 14]
```diff
@@ -0,0 +1,609 @@
...
+        self._create_gpu_table()
+        self._create_server_table()
+
+    def tabulate_results(self):
+        """
+        The function called at the end of all runs
+        FOR ALL MODELs that compiles all results and
+        dumps the data into tables for exporting.
+        """
+
```
Comment (nv-hwoo) [inline model_analyzer/result/result_table_manager.py:73]: Remove space.

File: model_analyzer/analyzer.py [hunk 15]
```diff
@@ -127,18 +128,20 @@ def analyze(self, mode, verbose):
...
         self._result_manager.compile_and_sort_results()
         self._report_manager.create_summaries()
         self._report_manager.export_summaries()
 
         # Dump to tables and write to disk
-        self._result_manager.tabulate_results()
-        self._result_manager.export_results()
+        self._result_table_manager.create_tables()
+        self._result_table_manager.tabulate_results()
+        self._result_table_manager.export_results()
```
Comment (nv-hwoo) [inline model_analyzer/analyzer.py:142]: Is there a reason why these are called separately? (e.g. cases where we create tables but not tabulate results). Maybe it would be cleaner to hide these steps inside the ResultTableManager.

File: tests/test_report_manager.py [hunk 16]
```diff
@@ -389,7 +361,7 @@ def _test_summary_counts(self, add_table_fn, add_plot_fn,
         avg_gpu_metrics = {0: {"gpu_used_memory": 6000, "gpu_utilization": 60}}
         for i in range(10):
             p99 = 20 + i
-            throughput = 100 - 10 * i if default_within_top else 100 + 10 * i
+            throughput = 100 - 10 * i if default_within_top else 200 + 10 * i
```
Comment (debermudez) [inline tests/test_report_manager.py:364]: This is beyond magic numbers. This is a whole spellbook 😛.  
Can we add a comment to explain this formula?

File: model_analyzer/result/result_table_manager.py [hunk 17]
```diff
@@ -54,6 +54,112 @@ def __init__(self, config, result_manager):
...
+                                        self._config.filename_model_gpu)
+        logger.info(f"Exporting inference metrics to {metrics_inference_path}")
+        logger.info(f"Exporting GPU metrics to {metrics_gpu_path}")
+        self._export_model_csv(
+            inference_writer=FileWriter(filename=metrics_inference_path),
+            gpu_metrics_writer=FileWriter(filename=metrics_gpu_path),
+            column_separator=',')
+
+    def _determine_table_headers(self):
+
```
Comment (nv-hwoo) [inline model_analyzer/result/result_table_manager.py:123]: Remove space.

File: tests/common/test_utils.py [hunk 18]
```diff
@@ -28,6 +35,58 @@
...
+        config = ConfigCommandAnalyze()
+    else:
+        config = ConfigCommandProfile()
+
+    cli = CLI()
+    cli.add_subcommand(cmd=subcommand, help="", config=config)
+    cli.parse()
+    mock_config.stop()
+    return config
+
```
Comment (nv-hwoo) [inline tests/common/test_utils.py:54]: Remove space.

File: model_analyzer/result/result_table_manager.py [hunk 19]
```diff
@@ -0,0 +1,609 @@
...
+            model_config._model_config.parameters
+            for model_config in model_configs
+        ]
+
+        return instance_groups, max_batch_sizes, dynamic_batchings, cpu_onlys, backend_parameters
+
+    def _tabulate_measurement(self, model_name, instance_groups,
+                              max_batch_sizes, dynamic_batchings,
+                              run_config_measurement, passes, cpu_onlys,
+                              backend_parameters):
```
Comment (nv-hwoo) [inline model_analyzer/result/result_table_manager.py:447]: Hard to parse. Can we list arguments line by line as in `_get_common_row_items`?

Comment (debermudez) [reply model_analyzer/result/result_table_manager.py:447]: The formatter dictates this.

File: model_analyzer/result/result_table_manager.py (hunk 4 above)
Comment (tgerdesnv) [reply model_analyzer/result/result_table_manager.py:385]: it would appear so. Removing the comment

File: tests/test_plot_manager.py (hunk 13 above)
Comment (tgerdesnv) [reply tests/test_plot_manager.py:18]: yup

File: tests/test_report_manager.py (hunk 16 above)
Comment (tgerdesnv) [reply tests/test_report_manager.py:364]: done

File: tests/test_result_manager.py (hunk 6 above)
Comment (tgerdesnv) [reply tests/test_result_manager.py:119]: fixed

**Review by nv-braf — DISMISSED** [review]
Looks good!

File: model_analyzer/result/result_table_manager.py (hunk 3 above)
Comment (debermudez) [reply model_analyzer/result/result_table_manager.py:337]: Thats fine. This would be a boyscout change at most then. Not necessary for this story.

File: tests/test_model_manager.py (hunk 9 above)
Comment (debermudez) [reply tests/test_model_manager.py:18]: Cool thanks

**Review by debermudez — APPROVED** [review]

**Review by nv-hwoo — DISMISSED** [review]

**Review by nv-hwoo — APPROVED** [review]
